### PR TITLE
flutter_tool: CLI cleanup related to enum-mapped options

### PR DESCRIPTION
**Pull request by kevmoo**
Added ParserExtension on ArgParser with addEnumOption

Added new HasHelpText class, implemented it on two enums
Moved Browser enum to new model
Used switch expressions a few places
Eliminated enum helpers that are no longer needed

### CODE REVIEW

File: packages/flutter_tools/lib/src/drive/web_driver_service.dart [hunk 1]
```diff
@@ -224,19 +223,46 @@ class WebDriverService extends DriverService {
 }
 
 /// A list of supported browsers.
-enum Browser {
-  /// Chrome on Android: https://developer.chrome.com/multidevice/android/overview
+enum Browser implements HasHelpText {
+  /// Chrome on Android: https://developer.chrome.com/docs/multidevice/android/
```
Comment (kevmoo) [inline packages/flutter_tools/lib/src/drive/web_driver_service.dart:227]: this doc was last updated in 2014!! – but at least it doesn't 404

File: packages/flutter_tools/lib/src/drive/web_driver_service.dart [hunk 2]
```diff
@@ -224,19 +223,46 @@ class WebDriverService extends DriverService {
...
+        Browser.androidChrome => 'Chrome on Android (see also "--android-emulator").',
+        Browser.chrome => 'Google Chrome on this computer (see also "--chrome-binary").',
+        Browser.edge => 'Microsoft Edge on this computer (Windows only).',
+        Browser.firefox => 'Mozilla Firefox on this computer.',
+        Browser.iosSafari => 'Apple Safari on an iOS device.',
+        Browser.safari => 'Apple Safari on this computer (macOS only).',
+      };
+
+  @override
+  String get cliName => switch (this) {
```
Comment (ditman) [inline packages/flutter_tools/lib/src/drive/web_driver_service.dart:256]: Shouldn't this be called `snakeCaseName`, so it's more clear to somebody who adds a new `Browser` that this returns the `snake-case` version of the `snakeCase` name?

File: packages/flutter_tools/lib/src/artifacts.dart [hunk 3]
```diff
@@ -713,7 +713,7 @@ class CachedArtifacts implements Artifacts {
         if (mode == BuildMode.debug || mode == null) {
           return _fileSystem.path.join(engineDir, platformName);
         }
-        final String suffix = mode != BuildMode.debug ? '-${snakeCase(getModeName(mode), '-')}' : '';
+        final String suffix = mode != BuildMode.debug ? '-${snakeCase(mode.name, '-')}' : '';
```
Comment (ditman) [inline packages/flutter_tools/lib/src/artifacts.dart:716]: Isn't this `BuildMode` a very good candidate to be Enum-ified? [Source](https://github.com/flutter/flutter/blob/ba8d100597a4b85bf1069da0120ec40d7fa0aae8/packages/flutter_tools/lib/src/build_info.dart#L380).

In all the code this PR is touching, it's being used as an "enum-like" thing :/

File: packages/flutter_tools/lib/src/base/utils.dart [hunk 4]
```diff
@@ -53,13 +53,6 @@ String snakeCaseToTitleCase(String snakeCaseString) {
 /// Return the plural of the given word (`cat(s)`).
 String pluralize(String word, int count) => count == 1 ? word : '${word}s';
 
-/// Return the name of an enum item.
-String getEnumName(dynamic enumItem) {
-  final String name = '$enumItem';
-  final int index = name.indexOf('.');
-  return index == -1 ? name : name.substring(index + 1);
-}
```
Comment (ditman) [inline packages/flutter_tools/lib/src/base/utils.dart:61]: Yikes! Good riddance!

File: packages/flutter_tools/lib/src/drive/web_driver_service.dart (hunk 2 above)
Comment (ditman) [reply packages/flutter_tools/lib/src/drive/web_driver_service.dart:256]: Or just use the `snakeCase(name, '-')` function so you don't need this `cliName` getter at all?

File: packages/flutter_tools/lib/src/drive/web_driver_service.dart [hunk 5]
```diff
@@ -247,102 +273,80 @@ Map<String, dynamic> getDesiredCapabilities(
...
-          ],
-          'perfLoggingPrefs': <String, String>{
-            'traceCategories':
-            'devtools.timeline,'
-            'v8,blink.console,benchmark,blink,'
-            'blink.user_timing',
+}) =>
+    switch (browser) {
+      Browser.chrome => <String, dynamic>{
+          'acceptInsecureCerts': true,
```
Comment (ditman) [inline packages/flutter_tools/lib/src/drive/web_driver_service.dart:279]: Arrow syntax is being abused here (at least at the top level)

https://dart.dev/guides/language/effective-dart/usage#consider-using--for-simple-members

File: packages/flutter_tools/lib/src/runner/parser_extension.dart [hunk 6]
```diff
@@ -0,0 +1,29 @@
...
+      help: help,
+      allowed: values.map((T e) => e.cliName),
+      allowedHelp: Map<String, String>.fromEntries(
+        values.map((T e) => MapEntry<String, String>(e.cliName, e.helpText)),
+      ),
+    );
+  }
+}
+
+abstract interface class HasHelpText implements Enum {
```
Comment (ditman) [inline packages/flutter_tools/lib/src/runner/parser_extension.dart:26]: I'd suggest `HasArgParserMetadata` or similar. Your `HasHelpText` enums have more than `helpText` on them :P

File: packages/flutter_tools/lib/src/runner/parser_extension.dart [hunk 7]
```diff
@@ -0,0 +1,29 @@
+// Copyright 2014 The Flutter Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style license that can be
+// found in the LICENSE file.
+
+import 'package:args/args.dart';
+
+extension ParserExtension on ArgParser {
```
Comment (ditman) [inline packages/flutter_tools/lib/src/runner/parser_extension.dart:7]: I'd remove this extension. You are writing much more boilerplate here than the code that you save by accepting an enum.

If this were something like:

```dart
void addEnumOption<T extends CoolEnum>(T value)
```

And compute the rest from `value`, I'd be onboard. But just to save a couple of calls to `map`, this seems excessive.

File: packages/flutter_tools/lib/src/runner/flutter_command.dart [hunk 8]
```diff
@@ -130,7 +125,7 @@ abstract final class FlutterOptions {
 }
 
 /// flutter command categories for usage.
-class FlutterCommandCategory {
+abstract final class FlutterCommandCategory {
```
Comment (ditman) [inline packages/flutter_tools/lib/src/runner/flutter_command.dart:128]: This one also smells like a new enum with a cool `.toString()` implementation :/

**Pull request closed without merging**